File: bot_engine/config/hot_reload.py
"""Hot-reload config subscriber via Redis Pub/Sub."""
from __future__ import annotations

import logging
import threading
import time
from typing import Callable

import redis

log = logging.getLogger(__name__)

CONFIG_CHANNEL = "customclaw:bot-config-changed"


def start_config_subscriber(
    redis_url: str,
    on_reload: Callable[[str], None],
) -> threading.Thread:
    """Start a daemon thread that subscribes to config change events.

    Uses a dedicated Redis connection for Pub/Sub (Redis requirement).
    Calls on_reload(bot_id) when a config change event is received.
    Reconnects automatically on connection loss.
    """

    def _listen() -> None:
        while True:
            try:
                r = redis.from_url(redis_url)
                pubsub = r.pubsub()
                pubsub.subscribe(CONFIG_CHANNEL)
                log.info(
                    "Config hot-reload subscriber started on channel: %s",
                    CONFIG_CHANNEL,
                )

                for message in pubsub.listen():
                    if message["type"] != "message":
                        continue
                    bot_id = message["data"]
                    if isinstance(bot_id, bytes):
                        bot_id = bot_id.decode()
                    log.info("Config reload event received for bot: %s", bot_id)
                    try:
                        on_reload(bot_id)
                    except Exception:
                        log.exception(
                            "Error during config hot-reload for bot: %s", bot_id
                        )
            except redis.ConnectionError:
                log.warning(
                    "Config subscriber lost Redis connection, reconnecting in 5s..."
                )
                time.sleep(5)
            except Exception:
                log.exception(
                    "Config subscriber unexpected error, restarting in 5s..."
                )
                time.sleep(5)

    thread = threading.Thread(target=_listen, name="config-hot-reload", daemon=True)
    thread.start()
    return thread


def publish_config_change(redis_client: redis.Redis, bot_id: str) -> None:
    """Publish a config change event. Called by bot management tools."""
    redis_client.publish(CONFIG_CHANNEL, bot_id)

File: go/internal/config/hotreload.go
package config

import (
	"context"
	"log/slog"
	"time"

	"github.com/redis/go-redis/v9"
)

// ConfigChannel is the Redis Pub/Sub channel used to broadcast bot config changes.
const ConfigChannel = "customclaw:bot-config-changed"

// StartConfigSubscriber launches a goroutine that listens for config change events
// on ConfigChannel. When a message arrives, onReload is called with the bot ID.
// The goroutine reconnects automatically on connection loss.
// It stops when ctx is cancelled.
func StartConfigSubscriber(ctx context.Context, redisURL string, onReload func(botID string)) {
	go func() {
		for {
			if err := ctx.Err(); err != nil {
				return
			}
			if err := runSubscribeLoop(ctx, redisURL, onReload); err != nil {
				if ctx.Err() != nil {
					return
				}
				slog.Warn("config subscriber error, reconnecting", "error", err, "delay", "5s")
			}
			select {
			case <-ctx.Done():
				return
			case <-time.After(5 * time.Second):
			}
		}
	}()
}

// runSubscribeLoop creates a dedicated Pub/Sub connection and processes messages
// until the context is cancelled or a connection error occurs.
func runSubscribeLoop(ctx context.Context, redisURL string, onReload func(botID string)) error {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return err
	}
	rdb := redis.NewClient(opts)
	defer func() { _ = rdb.Close() }()

	ps := rdb.Subscribe(ctx, ConfigChannel)
	defer func() { _ = ps.Close() }()

	slog.Info("config hot-reload subscriber started", "channel", ConfigChannel)

	ch := ps.Channel()
	for {
		select {
		case <-ctx.Done():
			return nil
		case msg, ok := <-ch:
			if !ok {
				return nil
			}
			botID := msg.Payload
			slog.Info("config reload event received", "bot_id", botID)
			func() {
				defer func() {
					if r := recover(); r != nil {
						slog.Error("panic in config reload handler", "bot_id", botID, "panic", r)
					}
				}()
				onReload(botID)
			}()
		}
	}
}

// PublishConfigChange notifies all subscribers that a bot's configuration has changed.
// This is called by bot management tools after updating the database.
func PublishConfigChange(ctx context.Context, rdb *redis.Client, botID string) error {
	return rdb.Publish(ctx, ConfigChannel, botID).Err()
}
